File: templates/segment-trees/normal/pointAssignRangeMax.py
from math import inf
class MaxSegTree:
    def __init__(self, arr):
        self.arr = list(arr)
        self.n = len(arr)
        size = 1
        while size < self.n:
            size <<= 1
        self.size = size
        tree = [-inf] * (2 * size)
        base = size
        for i, v in enumerate(arr):
            tree[base + i] = v
        for idx in range(size - 1, 0, -1):
            tree[idx] = max(tree[idx << 1], tree[(idx << 1) | 1])
        self.tree = tree

    def _queryHalfOpen(self, l, r):
        tree = self.tree
        l += self.size
        r += self.size
        ans = -inf
        while l < r:
            if l & 1:
                if tree[l] > ans: ans = tree[l]
                l += 1
            if r & 1:
                r -= 1
                if tree[r] > ans: ans = tree[r]
            l >>= 1
            r >>= 1
        return ans

    # 0 if l>r
    def queryMax(self, l, r):
        if l > r: return 0
        return self._queryHalfOpen(l, r + 1)

    def pointAssignAndMutateArray(self, index, newVal):
        self.arr[index] = newVal
        tree = self.tree
        pos = self.size + index
        tree[pos] = newVal
        pos >>= 1
        while pos:
            new = max(tree[pos << 1], tree[(pos << 1) | 1])
            if tree[pos] == new:
                break
            tree[pos] = new
            pos >>= 1

File: templates/segment-trees/normal/pointAssignRangeMax.cpp
// TEMPLATE BY ISHAANBUILDSTHINGS
#include <bits/stdc++.h>
using namespace std;
template <typename T>
struct MaxSegTree {
    static constexpr T NEG = numeric_limits<T>::min();
    int n, size;
    vector<T> arr, tree;

    // O(n) time, O(n) space
    MaxSegTree(const vector<T>& a) : n(a.size()), arr(a) {
        size = 1;
        while (size < n) size <<= 1;
        tree.assign(2 * size, NEG);
        for (int i = 0; i < n; i++) tree[size + i] = a[i];
        for (int i = size - 1; i >= 1; i--)
            tree[i] = max(tree[i << 1], tree[i << 1 | 1]);
    }

    // O(log n). max over [l, r), NEG if empty
    T _queryHalfOpen(int l, int r) const {
        T ans = NEG;
        for (l += size, r += size; l < r; l >>= 1, r >>= 1) {
            if (l & 1) ans = max(ans, tree[l++]);
            if (r & 1) ans = max(ans, tree[--r]);
        }
        return ans;
    }

    // O(log n). recompute ancestors of leaf index, stopping once nothing changes
    void _pullUp(int index) {
        for (int pos = (size + index) >> 1; pos; pos >>= 1) {
            T v = max(tree[pos << 1], tree[pos << 1 | 1]);
            if (tree[pos] == v) break;
            tree[pos] = v;
        }
    }

    // O(log n). max over [l, r] inclusive, 0 if l > r
    T queryMax(int l, int r) const {
        if (l > r) return 0;
        return _queryHalfOpen(l, r + 1);
    }

    // O(1)
    T pointGet(int index) const { return tree[size + index]; }

    // O(log n)
    void pointAssign(int index, T newVal) {
        tree[size + index] = newVal;
        _pullUp(index);
    }

    // O(log n)
    void pointAssignAndMutateArray(int index, T newVal) {
        arr[index] = newVal;
        pointAssign(index, newVal);
    }

    // O(log n)
    void pointChmax(int index, T val) {
        if (val <= tree[size + index]) return;
        pointAssign(index, val);
    }

    // O(log n)
    void pointChmin(int index, T val) {
        if (val >= tree[size + index]) return;
        pointAssign(index, val);
    }
};